src/sage/rings/species.py

```diff
@@ -0,0 +1,406 @@
+from sage.combinat.integer_vector import IntegerVectors
+from sage.misc.cachefunc import cached_method
+from sage.monoids.indexed_free_monoid import IndexedFreeAbelianMonoid
+from sage.structure.parent import Parent
+from sage.structure.element import parent
+from sage.structure.unique_representation import UniqueRepresentation
+from sage.rings.integer_ring import ZZ
+from sage.categories.sets_with_grading import SetsWithGrading
+from sage.categories.graded_algebras_with_basis import GradedAlgebrasWithBasis
+from sage.groups.perm_gps.permgroup import PermutationGroup, PermutationGroup_generic
+from sage.groups.perm_gps.permgroup_named import SymmetricGroup
+from sage.libs.gap.libgap import libgap
+from sage.combinat.free_module import CombinatorialFreeModule
+from sage.sets.set import Set
+from sage.rings.burnside import ConjugacyClassOfSubgroups, ElementCache, _is_conjugate
+
+GAP_FAIL = libgap.eval('fail')
+
+class ConjugacyClassOfDirectlyIndecomposableSubgroups(ConjugacyClassOfSubgroups):
```

Comment:
I am not sure:

* wouldn't it be better to make this completely independent of `burnside.py`?
* I don't see the value of inheriting from `ConjugacyClassOfSubgroups` - we are overriding most methods here.
* most importantly, I think we also want an additional parent class here, to handle the normalization of the subgroups (via an `_element_constructor_` that checks validity of the input and puts the group into a cache.

Reply:
Yes, I'll make it independent of `burnside.py`. It won't inherit from `ConjugacyClassOfSubgroups`.

I don't see the use of caching just the groups: we care about the group paired with an assignment of the domain to variables, so it would make more sense to cache instances of `AtomicSpeciesElement`.

Reply:
I disagree for two reasons:
* this class will handle, so to say, the part which is "difficult" to normalize.  `AtomicSpeciesElement` has no reason to know how a (conjugacy class of a) directly indecomposable group is stored.
* `AtomicSpeciesElement` may reuse the same `ConjugacyClassOfDirectlyIndecomposablePermutationGroup` for different instances.  Since we have unique factorization, there is no gain in caching the assignment of the domain to variables. I also guess that also any other client of `ConjugacyClassOfDirectlyIndecomposablePermutationGroup` (e.g., the Burnside ring) will want normalization and caching.

Reply:
The Burnside ring is not a client of `ConjugacyClassOfDirectlyIndecomposablePermutationGroup` because in that ring the generators are the conjugacy classes of a group, regardless of whether they are directly decomposable or not. But yes, it would be good to have normalization and caching there too, so I hope to factor the common functionality out into `ElementCache` and then the parent of `ConjugacyClassOfDirectlyIndecomposablePermutationGroup` and `ConjugacyClassOfSubgroups` can inherit from it.

Reply:
I agree.  Yes, that's what I meant - since factorization is so fast, we might also use it there.  But I don't have any (scientific) use cases for the Burnside ring currently.